test/source-sink-errors/mzcompose.py

```diff
@@ -133,22 +144,120 @@ def assert_recovery(self, c: Composition) -> None:
         )
 
 
-disruptions = [
-    Disruption(
+@dataclass
+class PgDisruption:
+    name: str
+    breakage: Callable
+    expected_error: str
+    fixage: Optional[Callable]
+
+    def run_test(self, c: Composition) -> None:
+        print(f"+++ Running disruption scenario {self.name}")
+        seed = random.randint(0, 256**4)
+
+        c.down(destroy_volumes=True)
+        c.up("testdrive", persistent=True)
+        c.start_and_wait_for_tcp(services=["postgres", "materialized", "storaged"])
+        c.wait_for_materialized()
+
+        with c.override(
+            Testdrive(
+                no_reset=True,
+                seed=seed,
+                entrypoint_extra=["--initial-backoff=1s", "--backoff-factor=0"],
+            )
+        ):
+            self.populate(c)
+            self.breakage(c, seed)
+            self.assert_error(c, self.expected_error)
+
+            if self.fixage:
+                self.fixage(c, seed)
+                self.assert_recovery(c)
+
+    def populate(self, c: Composition) -> None:
+        # Create a source and a sink
+        c.testdrive(
+            dedent(
+                """
+                > CREATE SECRET pgpass AS 'postgres'
+                > CREATE CONNECTION pg TO POSTGRES (
+                    HOST postgres,
+                    DATABASE postgres,
+                    USER postgres,
+                    PASSWORD SECRET pgpass
+                  )
+
+                $ postgres-execute connection=postgres://postgres:postgres@postgres
+                ALTER USER postgres WITH replication;
+                DROP SCHEMA IF EXISTS public CASCADE;
+                CREATE SCHEMA public;
+
+                DROP PUBLICATION IF EXISTS mz_source;
+                CREATE PUBLICATION mz_source FOR ALL TABLES;
+
+                CREATE TABLE pk_table (pk INTEGER PRIMARY KEY, f2 TEXT);
+                INSERT INTO pk_table VALUES (1, 'one');
+                ALTER TABLE pk_table REPLICA IDENTITY FULL;
+                INSERT INTO pk_table VALUES (2, 'two');
+                INSERT INTO pk_table VALUES (3, 'three');
+
+                > CREATE SOURCE "source1"
+                  FROM POSTGRES CONNECTION pg (PUBLICATION 'mz_source')
+                  FOR TABLES ("pk_table");
+
+                > SELECT * FROM pk_table ORDER BY pk ASC;
```

Comment:
You may wish to put a TODO here about the errors during snapshot, or open a follow-up ticket.